google/resource_storage_bucket_iam_policy_test.go

```diff
@@ -0,0 +1,182 @@
+			{
+				Role: "roles/editor",
+				Members: []string{
+					"user:admin@hashicorptest.com",
+				},
+			},
+			{
+				Role: "roles/viewer",
+				Members: []string{
+					"user:admin@hashicorptest.com",
+				},
+			},
+		},
+	}
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccGoogleStorageBucketIAMPolicyDestroy,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: testGoogleStorageBucketsIAMPolicyBasic(bucketName, policy1),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckGoogleStorageBucketIAMPolicy(bucketName, policy1),
+				),
+			},
+
+			resource.TestStep{
+				Config: testGoogleStorageBucketsIAMPolicyBasic(bucketName, policy2),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckGoogleStorageBucketIAMPolicy(bucketName, policy2),
+				),
+			},
+		},
+	})
+}
+
+func testAccCheckGoogleStorageBucketIAMPolicy(bucket string, policy *storagev1.Policy) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		rs, ok := s.RootModule().Resources[bucket]
+		if !ok {
+			return fmt.Errorf("Not found: %s", bucket)
+		}
+
+		if rs.Primary.ID == "" {
+			return fmt.Errorf("No ID is set")
+		}
+
+		config := testAccProvider.Meta().(*Config)
+
+		p, err := config.clientStorage.Buckets.GetIamPolicy(rs.Primary.ID).Do()
+
+		if err != nil {
+			return fmt.Errorf("Error retrieving contents of IAMPolicy for bucket %s: %s", bucket, err)
+		}
+
+		if !reflect.DeepEqual(p.Bindings, policy.Bindings) {
+			return fmt.Errorf("Incorrect iam policy bindings. Expected '%s', got '%s'", policy.Bindings, p.Bindings)
+		}
+
+		if _, ok = rs.Primary.Attributes["etag"]; !ok {
+			return fmt.Errorf("Etag should be set.")
+		}
+
+		if rs.Primary.Attributes["etag"] != p.Etag {
+			return fmt.Errorf("Incorrect etag value. Expected '%s', got '%s'", p.Etag, rs.Primary.Attributes["etag"])
+		}
+
+		return nil
+	}
+}
+
+func testAccGoogleStorageBucketIAMPolicyDestroy(s *terraform.State) error {
+	config := testAccProvider.Meta().(*Config)
+
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "google_storage_bucket_iam_policy" {
+			continue
+		}
+
+		bucket := rs.Primary.Attributes["bucket"]
+
+		policy, err := config.clientStorage.Buckets.GetIamPolicy(bucket).Do()
+
+		if err != nil && len(policy.Bindings) > 0 {
+			return fmt.Errorf("Bucket %s policy hasn't been deleted", bucket)
+		}
+	}
+
+	return nil
+}
+
+func testGoogleStorageBucketsIAMPolicyBasic(bucketName string, policy *storagev1.Policy) string {
+	var bindingBuffer bytes.Buffer
+
+	for _, binding := range policy.Bindings {
+		bindingBuffer.WriteString("binding {\n")
+		bindingBuffer.WriteString(fmt.Sprintf("role = \"%s\"\n", binding.Role))
+		bindingBuffer.WriteString(fmt.Sprintf("members = [\n"))
+		for _, member := range binding.Members {
+			bindingBuffer.WriteString(fmt.Sprintf("\"%s\",\n", member))
+		}
+		bindingBuffer.WriteString("]}\n")
```

Comment:
This part of the code scares me, to be honest.

Reply:
I don't understand Why?

Reply:
this is the same than resource_google_folder_iam_policy_test.go;

I have added Comment on source code

Reply:
Apologies, missed this response. It largely scares me because we're dynamically building the config string in a buffer, which is more complicated than tests usually require. I think this may be more generalized than is strictly necessary, but it's really not worth holding a merge over.